design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+
+- Set up alerting on Operator CI failures
+  - So we catch issues sooner.
+- Operator sets controller name in Contour ConfigMap
+  - To ensure Contour does not immediately exit on startup.
+- Operator stops setting any status on Gateway API resources
+  - To ensure Contour and Operator do not "fight" over status
+- Contour does not reject GatewayClasses if parameters ref is set
+  - At this point, we should have a workable intermediate solution for the moment and can cut a patch release
+  - Reconciliation will be duplicated but this should "work" for now
+  - User creates GatewayClass/Contour CRD with GatewayClass name
+  - Operator reconciles, does not set status
+  - User creates Gateway referencing GatewayClass
+  - Operator reconciles, creates a Contour
+  - Contour reconciles GatewayClass and Gateway, sets status, etc.
+- Operator deprecates Contour CRD GatewayClassRef field
+  - Currently the gateway reference has to be a GatewayClass name which is then validated to have the correct (hardcoded) controller string
+  - We should instead ensure the CRD has a field for setting the controller string the instance of Contour will watch for and reconcile GatewayClasses for
+  - The Operator should also set status to invalid on Contour objects that try to use a GatewayClass controller string that already exists, as it doesn't really make sense for multiple instances of Contour to watch for the same Controller string
+- Contour no longer requires Gateway namespace/name to be configured
+  - This is a planned deprecation anyway
+  - Will help with next step
+- Remove GatewayClass and Gateway controllers from Operator
+  - Planned deprecation anyways
+  - Users will no longer create a GatewayClass that references a Contour CRD to get a Contour that is configured for Gateway API.
+  - Instead, users will create a Contour CRD with a GatewayClass name and the operator will turn that into a Contour with the appropriate config.
+  - Operator will need to expand on Contour CRD controller, ensure Contour deployment resources are created when GatewayClass ref is set on Contour object
+- Move Operator into main Contour repository (long-term/optional)
+  - This option is a longer term project that would help us sort out process issues between Contour and Operator development.
+  - We would move the Operator codebase into the Contour one and deprecate/remove the Operator specific repo.
+  - Pros
+    - Allows us to test, get feedback in one repository, each PR will be tested against Contour, Contour integration in Operator
+    - Operator can be tested against specific Contour image versions, currently Operator in CI is just pointed at the `main` tag so we don't always know what is tested in a CI run
+    - Shared code for common functions, no need to extract libraries and sync versions across repos.
+  - Cons
+    - Doesn't follow typical operator/component repo split pattern
+    - Changes unrelated to operator will cause CI to run, expensive+time consuming
```

Comment:
Another remediation step

Seems we currently don't have a deployment guide of Contour Operator + Gateway API.  It would be helpful to have one under the Contour Operator repo representing the current state of design.  It's ok to keep updating the design.

Reply:
What's missing is what the user is expected to create when he wants to use the Operator to support Gateway API.  We've had a lot of design docs but nothing really user facing.  The current Operator quickstart for ex. gives that in the non-gateway API scenario, it just creates all the operator associated CRs in a single yaml and then a Contour + Envoy deployment in a second yaml apply.